### PR TITLE
geogfn: implement Covers using S2

### PR DESCRIPTION
**Pull request by otan**
This commit introduces covers using the primitives available in S2. Most
of the nitty gritty is in comments - but the main thing to note is that
we are introducing a "within 1cm" of accuracy (as intersects is very very
painful in a spheroid), which in geography terms should be good enough^TM
for any application and is line with what S2 provides for it's
primitives.

Note PostGIS has a 0.00001m (0.001cm) tolerance for ST_Intersects.

Release note: None

### CODE REVIEW

**Review by sumeerbhola — CHANGES_REQUESTED** [review]
> Reviewed 1 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47731)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geogfn/covers.go, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5TJXqgBRxHc6RCBR2M:-M5TJXqgBRxHc6RCBR2N:b-2m6kpe) ([raw file](https://github.com/cockroachdb/cockroach/blob/69cdf59abd24f8f9b60ce3988b8de0f739524faf/pkg/geo/geogfn/covers.go#L31)):*
> ```Go
> // This calculation is done on the sphere.
> func CoveredBy(a *geo.Geography, b *geo.Geography) (bool, error) {
> 	return doBinaryPredicate(b, a, &regionCoverPredicate{})
> ```

Is this decomposition valid?
According to the PostGIS documentation
"A MULTIPOLYGON is valid if and only if all of its elements are valid and the interiors of no two elements intersect. The boundaries of any two elements may touch, but only at a finite number of POINTs."
A line could pass from one polygon into another at this point where they touch.
___
*[pkg/geo/geogfn/covers.go, line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5THFiOCCFFgL10dJN3:-M5THFiOCCFFgL10dJN4:b-lqpnsq) ([raw file](https://github.com/cockroachdb/cockroach/blob/69cdf59abd24f8f9b60ce3988b8de0f739524faf/pkg/geo/geogfn/covers.go#L37)):*
> ```Go
> type regionCoverPredicate struct{}
> 
> var _ binaryPredicate = (*regionCoverPredicate)(nil)
> ```

will some implementations of `regionCoverPredicate` have internal state, which is why this is an interface and not a func type?
___
*[pkg/geo/geogfn/covers.go, line 86 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5TI0RQ34-G2AlewaNQ:-M5TI0RQ34-G2AlewaNR:bfjx3wd) ([raw file](https://github.com/cockroachdb/cockroach/blob/69cdf59abd24f8f9b60ce3988b8de0f739524faf/pkg/geo/geogfn/covers.go#L86)):*
> ```Go
> 
> // polylineCoversPointsWithIdx returns whether a polyline covers a given point.
> // If true, it will also return an index of which point there was an intersection.
> ```

index of the edge?
___
*[pkg/geo/geogfn/covers.go, line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5TIX813YsWWH1_WDvg:-M5TIX813YsWWH1_WDvh:b-wgqtme) ([raw file](https://github.com/cockroachdb/cockroach/blob/69cdf59abd24f8f9b60ce3988b8de0f739524faf/pkg/geo/geogfn/covers.go#L99)):*
> ```Go
> func polygonCoversPoint(a *s2.Polygon, b s2.Point) bool {
> 	// Account for the case where b is on the edge of the polygon.
> 	return a.ContainsPoint(b) || a.IntersectsCell(s2.CellFromPoint(b))
> ```

is the or with `IntersectsCell()` done to be consistent with the definition of covering for (polyline, point) pairs?
___
*[pkg/geo/geogfn/covers.go, line 122 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5TJ70p5ZDZhJ1bWFbi:-M5TJ70p5ZDZhJ1bWFbj:bkjvtx0) ([raw file](https://github.com/cockroachdb/cockroach/blob/69cdf59abd24f8f9b60ce3988b8de0f739524faf/pkg/geo/geogfn/covers.go#L122)):*
> ```Go
> // polylineCoversPolylineOrdered returns whether a polyline covers a polyline
> // in the same ordering.
> func polylineCoversPolylineOrdered(a *s2.Polyline, b *s2.Polyline) bool {
> ```

I need to reread this function.
___
*[pkg/geo/geogfn/geogfn.go, line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5SxL_N2-4xFjAcRCRx:-M5SxL_O2YsOD3ekwn3w:b-pqvb8e) ([raw file](https://github.com/cockroachdb/cockroach/blob/69cdf59abd24f8f9b60ce3988b8de0f739524faf/pkg/geo/geogfn/geogfn.go#L27)):*
> ```Go
> // doBinaryPredicate applies the binary predicate on a and b.
> // In collections, any single item in collection A has a binary predicate to be true
> // in all of collection B will return true.
> ```

This sentence seems to be missing a word or two. Perhaps this could be restated as:
- a and b are represented as sets of `s2.Region` objects, set-A and set-B.
- the binary predicate is true iff there exists an element e in set-A such that the binary predicate is true for all (e, x) where x is an element of set-B.

**Review by otan — COMMENTED** [review]
Going to need to think over this.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-:-M5TNR_sEEgmGUjZIdKb:bpmfpaf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geogfn/covers.go, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5TJXqgBRxHc6RCBR2M:-M5TLeSO7VenSofckUwk:b-v8i420) ([raw file](https://github.com/cockroachdb/cockroach/blob/69cdf59abd24f8f9b60ce3988b8de0f739524faf/pkg/geo/geogfn/covers.go#L31)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Is this decomposition valid?
According to the PostGIS documentation
"A MULTIPOLYGON is valid if and only if all of its elements are valid and the interiors of no two elements intersect. The boundaries of any two elements may touch, but only at a finite number of POINTs."
A line could pass from one polygon into another at this point where they touch.
</blockquote></details>

PostGIS won't return true for this. Interesting.

```
otan=# select st_covers('MULTIPOLYGON(((0.0 0.0, 0.0 1.0, 1.0 1.0, 1.0 0.0, 0.0 0.0)), ((1.0 0.0, 1.0 1.0, 2.0 2.0, 2.0 0.0, 1.0 0.0)))', 'LINESTRING(0.0 0.0, 2.0 0.0)');
 st_covers
-----------
 t
(1 row)

otan=# select st_covers('MULTIPOLYGON(((0.0 0.0, 0.0 1.0, 1.0 1.0, 1.0 0.0, 0.0 0.0)), ((1.0 0.0, 1.0 1.0, 2.0 2.0, 2.0 0.0, 1.0 0.0)))'::geography, 'LINESTRING(0.0 0.0, 2.0 0.0)'::geography);
 st_covers
-----------
 f
(1 row)
```
___
*[pkg/geo/geogfn/covers.go, line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5THFiOCCFFgL10dJN3:-M5TL_BQ4yc_XB0IRY1a:b3vinko) ([raw file](https://github.com/cockroachdb/cockroach/blob/69cdf59abd24f8f9b60ce3988b8de0f739524faf/pkg/geo/geogfn/covers.go#L37)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

will some implementations of `regionCoverPredicate` have internal state, which is why this is an interface and not a func type?
</blockquote></details>

Yes -- e.g. DWithin will need state (distance).
___
*[pkg/geo/geogfn/covers.go, line 86 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5TI0RQ34-G2AlewaNQ:-M5TM9xY2mCQMU2j7Dc3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/69cdf59abd24f8f9b60ce3988b8de0f739524faf/pkg/geo/geogfn/covers.go#L86)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

index of the edge?
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/covers.go, line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5TIX813YsWWH1_WDvg:-M5TMAOzDUcnTs_CXIw4:bb77ufg) ([raw file](https://github.com/cockroachdb/cockroach/blob/69cdf59abd24f8f9b60ce3988b8de0f739524faf/pkg/geo/geogfn/covers.go#L99)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

is the or with `IntersectsCell()` done to be consistent with the definition of covering for (polyline, point) pairs?
</blockquote></details>

Yes.

**Comment by otan** [comment]
looks like I'll need to revisit this...

**Comment by otan** [comment]
this is back and ready for another look.

**Review by sumeerbhola — CHANGES_REQUESTED** [review]
> Reviewed 1 of 4 files at r1, 2 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47731)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geogfn/covers.go, line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5hEj9PCwirW12eOW62:-M5hEj9PCwirW12eOW63:b-vsttz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/baa384e8cc1972781efda08da90171116f55ad25/pkg/geo/geogfn/covers.go#L24)):*
> ```Go
> // Due to minor inaccuracies and lack of certain primitives in S2,
> // precision for Covers will be for up to 1cm.
> // NOTE: LINESTRING/LINESTRING and currently inaccurate - in particular, if they
> ```

s/and/are/
also there seem to some words missing in "but reference different points may cover, but due"
___
*[pkg/geo/geogfn/covers.go, line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5hEzXtE0cW1O7lsT4B:-M5hEzXtE0cW1O7lsT4C:b-iiq4y9) ([raw file](https://github.com/cockroachdb/cockroach/blob/baa384e8cc1972781efda08da90171116f55ad25/pkg/geo/geogfn/covers.go#L26)):*
> ```Go
> // NOTE: LINESTRING/LINESTRING and currently inaccurate - in particular, if they
> // are on top of each other but reference different points may cover, but due to
> // floating point precision may miss. This is in line with PostGIS experiments.
> ```

Can you elaborate on "in line ..." -- do you mean PostGIS is similarly inaccurate?
Can you add an example here?
___
*[pkg/geo/geogfn/covers.go, line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5hFF4Y8sgjefbmvsOb:-M5hFF4Y8sgjefbmvsOc:bbss8e0) ([raw file](https://github.com/cockroachdb/cockroach/blob/baa384e8cc1972781efda08da90171116f55ad25/pkg/geo/geogfn/covers.go#L30)):*
> ```Go
> // touches the boundary, we will return false despite needing to return true.
> // Furthermore, LINESTRINGS that are covered in the MULTIPOLYGON case are not
> // yet considered. This is in line with PostGIS experiments.
> ```

There are two sentences preceding the "in line with PostGIS" -- does this apply to both?
Can you revise the comment to make a bulleted list of issues and how each compares with PostGIS?
___
*[pkg/geo/geogfn/covers.go, line 45 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5hFo464fmP0LrvTpXK:-M5hFo464fmP0LrvTpXL:bv9en57) ([raw file](https://github.com/cockroachdb/cockroach/blob/baa384e8cc1972781efda08da90171116f55ad25/pkg/geo/geogfn/covers.go#L45)):*
> ```Go
> 	}
> 	// Check whether at least one of regions in A covers at least one of each
> 	// of regions in B.
> ```

This comment reads awkwardly, perhaps because it tries to follow the code structure, and I could not figure out what it meant, while the code is straightforward to understand.
How about:
Check whether each region in B is covered by at least one region in A.
___
*[pkg/geo/geogfn/covers.go, line 77 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5hGYJa4D0NK7KU45-8:-M5hGYJa4D0NK7KU45-9:b-7i9en6) ([raw file](https://github.com/cockroachdb/cockroach/blob/baa384e8cc1972781efda08da90171116f55ad25/pkg/geo/geogfn/covers.go#L77)):*
> ```Go
> 			return aRegion.ContainsPoint(bRegion), nil
> 		case *s2.Polyline:
> 			return false, nil
> ```

are polylines where all the points are identical allowed?
Same question for polygon.
Can you add code comments related to these corner cases?
___
*[pkg/geo/geogfn/covers.go, line 89 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5hGunc7VB8PDcCJp2Y:-M5hGunc7VB8PDcCJp2Z:bz7uwo9) ([raw file](https://github.com/cockroachdb/cockroach/blob/baa384e8cc1972781efda08da90171116f55ad25/pkg/geo/geogfn/covers.go#L89)):*
> ```Go
> 		case *s2.Polyline:
> 			return polylineCoversPolyline(aRegion, bRegion), nil
> 		case *s2.Polygon:
> ```

can one have a polygon that degenerates to a polyline?
___
*[pkg/geo/geogfn/covers.go, line 155 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5hIiFz3aGUPPry4jNM:-M5hIiFz3aGUPPry4jNN:b-ot7ze8) ([raw file](https://github.com/cockroachdb/cockroach/blob/baa384e8cc1972781efda08da90171116f55ad25/pkg/geo/geogfn/covers.go#L155)):*
> ```Go
> func polylineCoversPolylineOrdered(a *s2.Polyline, b *s2.Polyline) bool {
> 	aCoversStartOfB, aCoverBStart := polylineCoversPointWithIdx(a, (*b)[0])
> 	// We must first check that the start and end of B is contained by A.
> ```

the comment says "start and end of B" but I only see "start of B" here.
___
*[pkg/geo/geogfn/covers.go, line 160 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5hIr3q4QurE8tOaVi0:-M5hIr3q4QurE8tOaVi1:b-isrix3) ([raw file](https://github.com/cockroachdb/cockroach/blob/baa384e8cc1972781efda08da90171116f55ad25/pkg/geo/geogfn/covers.go#L160)):*
> ```
> is the first edge that polyline "B".
> ```
 sentence is incomplete.
___
*[pkg/geo/geogfn/covers.go, line 267 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5hI5F733ZizWxV4ogc:-M5hI5F733ZizWxV4ogd:b-yoma6r) ([raw file](https://github.com/cockroachdb/cockroach/blob/baa384e8cc1972781efda08da90171116f55ad25/pkg/geo/geogfn/covers.go#L267)):*
> ```Go
> func polygonCoversPolygon(a *s2.Polygon, b *s2.Polygon) bool {
> 	// We can rely on Contains here, as if the exteriors are the shared it will
> 	// still return true in here.
> ```

I didn't understand this sentence ("are the shared", "true in here") -- can you rephrase or elaborate?

**Review by otan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47731)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/geo/geogfn/covers.go, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5TJXqgBRxHc6RCBR2M:-M5hOSxLBnRU31LL8Up2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/69cdf59abd24f8f9b60ce3988b8de0f739524faf/pkg/geo/geogfn/covers.go#L31)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

PostGIS won't return true for this. Interesting.

```
otan=# select st_covers('MULTIPOLYGON(((0.0 0.0, 0.0 1.0, 1.0 1.0, 1.0 0.0, 0.0 0.0)), ((1.0 0.0, 1.0 1.0, 2.0 2.0, 2.0 0.0, 1.0 0.0)))', 'LINESTRING(0.0 0.0, 2.0 0.0)');
 st_covers
-----------
 t
(1 row)

otan=# select st_covers('MULTIPOLYGON(((0.0 0.0, 0.0 1.0, 1.0 1.0, 1.0 0.0, 0.0 0.0)), ((1.0 0.0, 1.0 1.0, 2.0 2.0, 2.0 0.0, 1.0 0.0)))'::geography, 'LINESTRING(0.0 0.0, 2.0 0.0)'::geography);
 st_covers
-----------
 f
(1 row)
```
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/covers.go, line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5THFiOCCFFgL10dJN3:-M5hOU8x5pRtzxpNZqg6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/69cdf59abd24f8f9b60ce3988b8de0f739524faf/pkg/geo/geogfn/covers.go#L37)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Yes -- e.g. DWithin will need state (distance).
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/covers.go, line 122 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5TJ70p5ZDZhJ1bWFbi:-M5hQoPL77nJrysdMTQj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/69cdf59abd24f8f9b60ce3988b8de0f739524faf/pkg/geo/geogfn/covers.go#L122)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I need to reread this function.
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/covers.go, line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5hEj9PCwirW12eOW62:-M5hPPqQC0MZzHwAZDfW:bj6stwb) ([raw file](https://github.com/cockroachdb/cockroach/blob/baa384e8cc1972781efda08da90171116f55ad25/pkg/geo/geogfn/covers.go#L24)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

s/and/are/
also there seem to some words missing in "but reference different points may cover, but due"
</blockquote></details>

I actually fixed this so removing it.
___
*[pkg/geo/geogfn/covers.go, line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5hEzXtE0cW1O7lsT4B:-M5hQPEr9fYah4AgyY-c:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/baa384e8cc1972781efda08da90171116f55ad25/pkg/geo/geogfn/covers.go#L26)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Can you elaborate on "in line ..." -- do you mean PostGIS is similarly inaccurate?
Can you add an example here?
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/covers.go, line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5hFF4Y8sgjefbmvsOb:-M5hQPb8DUPeD8KchS4W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/baa384e8cc1972781efda08da90171116f55ad25/pkg/geo/geogfn/covers.go#L30)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

There are two sentences preceding the "in line with PostGIS" -- does this apply to both?
Can you revise the comment to make a bulleted list of issues and how each compares with PostGIS?
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/covers.go, line 45 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5hFo464fmP0LrvTpXK:-M5hQa483WyWJXBv1Hvj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/baa384e8cc1972781efda08da90171116f55ad25/pkg/geo/geogfn/covers.go#L45)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This comment reads awkwardly, perhaps because it tries to follow the code structure, and I could not figure out what it meant, while the code is straightforward to understand.
How about:
Check whether each region in B is covered by at least one region in A.
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/covers.go, line 77 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5hGYJa4D0NK7KU45-8:-M5hQc7BDqP3JIx25HpD:b-bff5gm) ([raw file](https://github.com/cockroachdb/cockroach/blob/baa384e8cc1972781efda08da90171116f55ad25/pkg/geo/geogfn/covers.go#L77)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

are polylines where all the points are identical allowed?
Same question for polygon.
Can you add code comments related to these corner cases?
</blockquote></details>

It's assumed that all polylines are valid, and this is invalid. I can add a comment if you want but it's always inferred as soon as it is parsed from the WKT/WKB.
___
*[pkg/geo/geogfn/covers.go, line 89 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5hGunc7VB8PDcCJp2Y:-M5hQg9Q9CzNwjasDyqd:b-n0vj7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/baa384e8cc1972781efda08da90171116f55ad25/pkg/geo/geogfn/covers.go#L89)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

can one have a polygon that degenerates to a polyline?
</blockquote></details>

It's assumed that all polylines are valid, and this is invalid.
___
*[pkg/geo/geogfn/covers.go, line 155 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5hIiFz3aGUPPry4jNM:-M5hQr8Q8Pi2L-fsXKxO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/baa384e8cc1972781efda08da90171116f55ad25/pkg/geo/geogfn/covers.go#L155)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

the comment says "start and end of B" but I only see "start of B" here.
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/covers.go, line 160 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5hIr3q4QurE8tOaVi0:-M5hQvdh6Ds9oEMAXsPk:bwo33kj) ([raw file](https://github.com/cockroachdb/cockroach/blob/baa384e8cc1972781efda08da90171116f55ad25/pkg/geo/geogfn/covers.go#L160)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

> ```
> is the first edge that polyline "B".
> ```
 sentence is incomplete.
</blockquote></details>

is it?  added a f ew more words anyway
___
*[pkg/geo/geogfn/covers.go, line 267 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5hI5F733ZizWxV4ogc:-M5hRJeA841bBwGsfTsP:b-qf7lbl) ([raw file](https://github.com/cockroachdb/cockroach/blob/baa384e8cc1972781efda08da90171116f55ad25/pkg/geo/geogfn/covers.go#L267)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I didn't understand this sentence ("are the shared", "true in here") -- can you rephrase or elaborate?
</blockquote></details>

How is that for you?
___
*[pkg/geo/geogfn/geogfn.go, line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5SxL_N2-4xFjAcRCRx:-M5hOLuh2exJO6s6iQkC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/69cdf59abd24f8f9b60ce3988b8de0f739524faf/pkg/geo/geogfn/geogfn.go#L27)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This sentence seems to be missing a word or two. Perhaps this could be restated as:
- a and b are represented as sets of `s2.Region` objects, set-A and set-B.
- the binary predicate is true iff there exists an element e in set-A such that the binary predicate is true for all (e, x) where x is an element of set-B. 
</blockquote></details>

Done.

**Review by sumeerbhola — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47731)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geogfn/covers.go, line 89 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5hGunc7VB8PDcCJp2Y:-M5xfn4wDLMbtUmpdObq:bebf4j7) ([raw file](https://github.com/cockroachdb/cockroach/blob/baa384e8cc1972781efda08da90171116f55ad25/pkg/geo/geogfn/covers.go#L89)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

It's assumed that all polylines are valid, and this is invalid.
</blockquote></details>

did you mean "all polygons are valid"?
___
*[pkg/geo/geogfn/covers.go, line 160 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5hIr3q4QurE8tOaVi0:-M5xg1ou10sTYtDyZWlq:b-26jzub) ([raw file](https://github.com/cockroachdb/cockroach/blob/baa384e8cc1972781efda08da90171116f55ad25/pkg/geo/geogfn/covers.go#L160)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

is it?  added a f ew more words anyway
</blockquote></details>

"is the first that polyine "B"" ends abruptly. that what?
I was expecting something like:
We have found "aIdx" which is the first edge in polyline A that includes the starting vertex of polyline "B". Start checking the covering from this edge.

Aren't LINESTRINGS allowed to be non-simple, if they cross themselves? In that case don't we need to restart a failed search from the next line in A that covers the starting vertex of B?
___
*[pkg/geo/geogfn/covers.go, line 36 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5xfULt3ed23ki6Rj49:-M5xfULu-corTFBL6Dke:bo3okt4) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6057b541708c18dcaa6b3ef287e5ff48b40a400/pkg/geo/geogfn/covers.go#L36)):*
> ```Go
> //
> // * Furthermore, LINESTRINGS that are covered in multiple POLYGONs inside
> //   MULTIPOLYGON case are not but NOT within a single POLYGON in the MULTIPOLYGON
> ```

s/case are not//
___
*[pkg/geo/geogfn/covers.go, line 52 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5xfeNd9y79S-SM78n1:-M5xfeNd9y79S-SM78n2:bwwuzm3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6057b541708c18dcaa6b3ef287e5ff48b40a400/pkg/geo/geogfn/covers.go#L52)):*
> ```Go
> 	}
> 
> 	// We need to check each region in B is covered in at least
> ```

nit: is covered by ...

**Review by otan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47731)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/geo/geogfn/covers.go, line 89 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5hGunc7VB8PDcCJp2Y:-M5yGfaCE1JZJCk4ktv9:b-sde195) ([raw file](https://github.com/cockroachdb/cockroach/blob/baa384e8cc1972781efda08da90171116f55ad25/pkg/geo/geogfn/covers.go#L89)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

did you mean "all polygons are valid"?
</blockquote></details>

ah yeah.
___
*[pkg/geo/geogfn/covers.go, line 160 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5hIr3q4QurE8tOaVi0:-M5yGDIODDgyEirf9nZO:b-vgc1us) ([raw file](https://github.com/cockroachdb/cockroach/blob/baa384e8cc1972781efda08da90171116f55ad25/pkg/geo/geogfn/covers.go#L160)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

"is the first that polyine "B"" ends abruptly. that what?
I was expecting something like:
We have found "aIdx" which is the first edge in polyline A that includes the starting vertex of polyline "B". Start checking the covering from this edge.

Aren't LINESTRINGS allowed to be non-simple, if they cross themselves? In that case don't we need to restart a failed search from the next line in A that covers the starting vertex of B?
</blockquote></details>

Ah woops!

Polylines can't cross themselves (https://postgis.net/docs/using_postgis_dbmanagement.html#OGC_Validity).
___
*[pkg/geo/geogfn/covers.go, line 36 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-M5xfULt3ed23ki6Rj49:-M5yGalc49luf4po67eR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6057b541708c18dcaa6b3ef287e5ff48b40a400/pkg/geo/geogfn/covers.go#L36)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

s/case are not//
</blockquote></details>

Done.

**Review by sumeerbhola — COMMENTED** [review]
:lgtm: 

> Reviewed 1 of 4 files at r2, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47731#-:-M60qa9A1XGo5GQS41el:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @sumeerbhola)

**Comment by otan** [comment]
TFTR!

bors r=sumeerbhola
